### PR TITLE
Make ViewComponentTagHelper use extension intermediate nodes

### PR DESCRIPTION
**Pull request by ajaybhargavb**
#1368 

@NTaylorMullen @rynowak 

Copied the same changes to the Version1_X project. It's better to look at each commit separately.

### CODE REVIEW

File: src/Microsoft.AspNetCore.Mvc.Razor.Extensions.Version1_X/ViewComponentTagHelperIntermediateNode.cs [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+using System;
+using Microsoft.AspNetCore.Razor.Language;
+using Microsoft.AspNetCore.Razor.Language.CodeGeneration;
+using Microsoft.AspNetCore.Razor.Language.Intermediate;
+
+namespace Microsoft.AspNetCore.Mvc.Razor.Extensions.Version1_X
+{
+    public sealed class ViewComponentTagHelperIntermediateNode : ExtensionIntermediateNode
+    {
+        public override IntermediateNodeCollection Children { get; } = new IntermediateNodeCollection();
```
Comment (NTaylorMullen) [inline src/Microsoft.AspNetCore.Mvc.Razor.Extensions.Version1_X/ViewComponentTagHelperIntermediateNode.cs]: Will there ever be children?

File: src/Microsoft.AspNetCore.Mvc.Razor.Extensions/ViewComponentTagHelperTargetExtension.cs [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+using Microsoft.AspNetCore.Razor.Language;
+using Microsoft.AspNetCore.Razor.Language.CodeGeneration;
+
+namespace Microsoft.AspNetCore.Mvc.Razor.Extensions
+{
+    internal class ViewComponentTagHelperTargetExtension : IViewComponentTagHelperTargetExtension
+    {
+        private static readonly string[] PublicModifiers = new[] { "public" };
+
+        public string TagHelperTypeName { get; set; } = "Microsoft.AspNetCore.Razor.TagHelpers.TagHelper";
```
Comment (NTaylorMullen) [inline src/Microsoft.AspNetCore.Mvc.Razor.Extensions/ViewComponentTagHelperTargetExtension.cs]: Holy configuration. At least its internal lol

File: src/Microsoft.AspNetCore.Mvc.Razor.Extensions.Version1_X/ViewComponentTagHelperTargetExtension.cs [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+        {
+            var propertyNames = tagHelper.BoundAttributes.Select(attribute => attribute.GetPropertyName());
+            var joinedPropertyNames = string.Join(", ", propertyNames);
+            var parametersString = $"new {{ { joinedPropertyNames } }}";
+            var viewComponentName = tagHelper.GetViewComponentName();
+            var methodParameters = new[] { $"\"{viewComponentName}\"", parametersString };
+            return methodParameters;
+        }
+
+        private void BuildTargetElementString(CodeWriter writer, TagHelperDescriptor tagHelper)
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.Razor.Extensions.Version1_X/ViewComponentTagHelperTargetExtension.cs]: Can we rename these from `Build` to `Writer`?

File: src/Microsoft.AspNetCore.Mvc.Razor.Extensions.Version1_X/ViewComponentTagHelperIntermediateNode.cs (hunk 1 above)
Comment (ajaybhargavb) [inline src/Microsoft.AspNetCore.Mvc.Razor.Extensions.Version1_X/ViewComponentTagHelperIntermediateNode.cs]: I don't think so.

**Pull request merged**